### ISSUE TITLE
TCP traffic appears to be load balanced between all running containers

### ISSUE THREAD

**Issue by d3v1an7** [containous/traefik#4930]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [x] No

### What did you do?

Put together a minimal config to test TCP functionality with Redis (see below). Decided a ping test would be the best way to see if it was working to start with.

**Testing method**
```for i in `seq 1 1000`; do echo 'ping' | nc localhost 6379; done | uniq -c```

### What did you expect to see?

```1000 +PONG```
(ping responded with pong for 100% of requests)

### What did you see instead?

```390 +PONG```
(ping responded with pong for 30-50% of requests)

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.0-alpha4
Codename:     faisselle
Go version:   go1.12.4
Built:        2019-04-17T11:38:21Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

`traefik/config/traefik.toml`

```toml
[api]

[entrypoints]
  [entrypoints.traefik]
    address = ":8880"
  [entrypoints.redis-port]
    address = ":6379"

[global]
  debug = true

[log]
  format = "json"
  logLevel = "debug"

[providers]
    [providers.docker]
      defaultRule = "Host(`{{ normalize .Name }}.docker.localhost`)"
      endpoint = "unix:///var/run/docker.sock"
      watch = true
```

`traefik/docker-compose.yaml`

```yaml
version: '3'
services:
  proxy:
    container_name: reverse-proxy
    image: traefik:v2.0-alpine
    networks:
      - shared_infrastructure
    ports:
      - "6379:6379"
      - "8880:8880"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./config/traefik.toml:/etc/traefik/traefik.toml:ro
networks:
  shared_infrastructure:
    external: true
```

`redis/docker-compose.yaml`
```yaml
version: '3'
services:
  redis:
    container_name: redis
    image: redis:4-alpine
    networks:
      - shared_infrastructure
    labels:
      - "traefik.docker.network=shared_infrastructure"
      - "traefik.tcp.routers.redis.entrypoints=redis-port"
      - "traefik.tcp.routers.redis.rule=HostSNI(`*`)"
      - "traefik.tcp.services.redis.loadbalancer.server.port=6379"
    ports:
      - "6379"
    restart: always
networks:
  shared_infrastructure:
    external: true
```

`/api/rawdata`

```json
{
  "docker": {
    "HTTP": {
      "routers": {
        "proxy_traefik": {
          "entryPoints": null,
          "service": "proxy_traefik",
          "rule": "Host(`proxy-traefik.docker.localhost`)"
        }
      },
      "services": {
        "proxy_traefik": {
          "loadbalancer": {
            "servers": [
              {
                "url": "http://172.18.0.3:80",
                "weight": 1
              }
            ],
            "method": "wrr",
            "passHostHeader": true
          }
        }
      }
    },
    "TCP": {
      "routers": {
        "redis": {
          "entryPoints": [
            "redis-port"
          ],
          "service": "redis_redis",
          "rule": "HostSNI(`*`)"
        }
      },
      "services": {
        "redis_redis": {
          "loadbalancer": {
            "servers": [
              {
                "address": "172.18.0.2:6379",
                "weight": 1
              }
            ],
            "method": "wrr"
          }
        }
      }
    },
    "TLSOptions": null,
    "TLSStores": null
  }
}
```

**Comment by dduportal** [comment]
Hi @d3v1an7 , thanks for your detailed feedback.

At first try, we are able to reproduce the behaviour you describe only on Docker4Mac. On Linux Manjaro with a local Docker Engine, the provided shell command does not work, but after adapting it, 100% of the ping are "ponged".

Could you try (as we are also doing) to do the same check but with the `redis client` instead of `nc`,
as macOS's `nc` command could give us a false positive here?

We'll let you know what we found on our side.

**Comment by dduportal** [comment]
Hi @d3v1an7 , more results of our further investigations:

- On docker4mac, using `redis-cli` (from `brew install redis`), there is no balancing issue:

```shell
for i in $(seq 1 1000); do redis-cli ping; done | grep -c PONG
1000
```

- On docker4mac, with the Redis container directly exposed, your script shows 100 % success on the pongs:

```yml
# redis/docker-compose.yaml
services:
  redis:
    ports:
      - "16379:6379"
    ...
```

```shell
for i in `seq 1 1000`; do echo 'ping' | nc localhost 16379; done | uniq -c
1000 +PONG
```

- On macOS, using the docker of the latest minikube VM, the nc command is stuck and never returns
any PONG at all

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of tidying up the issues.

I have to close this one because of its lack of activity :disappointed:

Feel free to re-open it or join our [Community Forum](https://community.containo.us/).

**Comment by dduportal** [comment]
I've put a reproduction case with `2.0.0-rc1`.

For later readers: redis **works** when using a redis-client! The issue here is only related to the `nc` command used as a client with Traefik.

- docker-compose.yml:

```yml
version: '3'

services:
  proxy:
    image: traefik:v2.0.0-rc1-alpine
    command:
      - --entrypoints.redis.address=:6379
      - --api
      - --providers.docker
    ports:
      - "6379:6379"
      - "8880:8880"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock

  redis:
    image: redis:4-alpine
    labels:
      - "traefik.tcp.routers.redis.entrypoints=redis"
      - "traefik.tcp.routers.redis.rule=HostSNI(`*`)"
      - "traefik.tcp.services.redis.loadbalancer.server.port=6379" # Redis Container exposes more than 1 port
    ports:
      - "16379:6379" # Exposed for comparing with/without Traefik
```

- Reproduction steps:

```shell
$ docker-compose up -d
# wait a few seconds for the stack to be ready
$ docker-compose exec proxy sh
/ # GATEWAY="$(route | grep default | awk '{print $2}')"
/ # echo $GATEWAY # Test the gateway value
...
/ # echo 'ping' | nc ${GATEWAY} 16379
+PONG
/ # for i in `seq 1 1000`; do echo 'ping' | nc $GATEWAY 16379; done | uniq -c
   1000 +PONG
/ # for i in `seq 1 1000`; do echo 'ping' | nc $GATEWAY 16379; done | uniq -c
   1000 +PONG
/ # for i in `seq 1 1000`; do echo 'ping' | nc $GATEWAY 6379; done | uniq -c
    390 +PONG
/ # for i in `seq 1 1000`; do echo 'ping' | nc $GATEWAY 6379; done | uniq -c
    345 +PONG
```

**Comment by traefiker** [comment]
Closed by #5366.
